omc3/kmod/analysis.py

```diff
@@ -247,41 +254,147 @@ def return_df(magnet1_df, magnet2_df, plane):
     elif check_polarity(magnet1_df, magnet2_df, -sign[plane]):
         return magnet2_df, magnet1_df
 
+      
+def get_BPM(kmod_input_params):
+
+    # listing the BPMs of the last quadrupole BPMs
+    BPM_dict = {
+        "IP1" : ["BPMSW.1L1", "BPMSW.1R1"], 
+        "IP2" : ["BPMSW.1L2", "BPMSW.1R2"], 
+        "IP3" : ["BPMW.4L3", "BPMW.4R3"], 
+        "IP4" : ["BPMWA.A5L4", "BPMWA.A5R4"], 
+        "IP5" : ["BPMSW.1L5", "BPMSW.1R5"],
+        "IP6" : ["BPMSE.4L6", "BPMSA.4R6"], 
+        "IP7" : ["BPMW.4L7", "BPMW.4R7"], 
+        "IP8" : ["BPMSW.1L8", "BPMSW.1R8"], 
+    }
+
+    # Selecting BPMs
+    BPML = BPM_dict[kmod_input_params.ip.upper()][0] + '.' + kmod_input_params.beam.upper()
+    BPMR = BPM_dict[kmod_input_params.ip.upper()][1] + '.' + kmod_input_params.beam.upper()
+
+    return BPML, BPMR
+
+def get_BPM_distance(kmod_input_params,BPML,BPMR):
+
+    model_filename = 'twiss_' + kmod_input_params.beam.lower() + '.dat'
+
+    # Position s of BPML
+    if (os.path.exists(os.path.join(f'{kmod_input_params.twiss_model_dir}', f'{model_filename}'))): # distance between actual BPMs if model is provided (more precise)
+        twiss_df = tfs.read(os.path.join(f'{kmod_input_params.twiss_model_dir}', f'{model_filename}'), index='NAME')
+        pos_1L = twiss_df.loc[BPML,'S']
+        pos_1R = twiss_df.loc[BPMR,'S']
+        BPM_distance = np.abs(pos_1R - pos_1L)/2.0
+    else:
+        BPM_distance = 22.965 # Distance between last quadrupole and IP (LSTAR) (less precise)
+
+    return BPM_distance
+
+def get_phase_from_model(kmod_input_params,plane):        # get phase from twiss model
+    model_filename = 'twiss_' + kmod_input_params.beam + '.dat'
```

Comment:
where do you get twiss_B1/2.dat?

Reply:
twiss_B1/B2 currently must be in --model_dir. We could think a better way.